ledger/acctonline.go

```diff
@@ -581,89 +620,105 @@ func (ao *onlineAccounts) lookupOnlineAccountData(rnd basics.Round, addr basics.
-		// 	return macct.accountData.GetOnlineAccountData(rewardsProto, rewardsLevel), nil
-		// }
+	if macct, has := ao.onlineAccountsCache.read(addr, rnd); has {
+		return macct.GetOnlineAccountData(rewardsProto, rewardsLevel), nil
+	}
 
-		ao.accountsMu.RUnlock()
-		needUnlock = false
-
-		// No updates of this account in the in-memory deltas; use on-disk DB.
-		// The check in roundOffset() made sure the round is exactly the one
-		// present in the on-disk DB.  As an optimization, we avoid creating
-		// a separate transaction here, and directly use a prepared SQL query
-		// against the database.
-		persistedData, err = ao.accountsq.lookupOnline(addr, rnd)
-		if persistedData.round == currentDbRound {
-			if persistedData.rowid != 0 {
-				// if we read actual data return it
-				// TODO: add another cache
-				// ao.baseOnlineAccounts.writePending(persistedData)
-				return persistedData.accountData.GetOnlineAccountData(rewardsProto, rewardsLevel), err
-			}
-			// otherwise return empty
-			return ledgercore.OnlineAccountData{}, err
-		}
+	ao.accountsMu.RUnlock()
+	needUnlock = false
 
-		if persistedData.round < currentDbRound {
-			ao.log.Errorf("accountUpdates.lookupOnlineAccountData: database round %d is behind in-memory round %d", persistedData.round, currentDbRound)
-			return ledgercore.OnlineAccountData{}, &StaleDatabaseRoundError{databaseRound: persistedData.round, memoryRound: currentDbRound}
+	// No updates of this account in the in-memory deltas; use on-disk DB.
+	// As an optimization, we avoid creating
+	// a separate transaction here, and directly use a prepared SQL query
+	// against the database.
+	persistedData, err = ao.accountsq.lookupOnline(addr, rnd)
+	if err != nil {
+		return ledgercore.OnlineAccountData{}, err
+	}
+
+	if persistedData.rowid != 0 { // if we read actual data return it
+		// lookupOnlineHistory does not fetch the account db round because of the following observation:
+		// 1. ao.onlineAccountsCache update happens with ao.accountsMu taken below and in postCommit
+		// 2. If we started reading the history (lookupOnlineHistory)
+		//   * before commitRound or while it is running => OK, read what is in DB and then add new entries in postCommit
+		//     - if commitRound deletes some history after, the cache has additional entries and updRound comparison gets a right value
+		//   * after commitRound but before postCommit => OK, read full history, ignore the update from postCommit in writeFront's updRound comparison
+		//   * after postCommit => OK, postCommit does not add new entry with writeFrontIfExist, but here all the full history is loaded
+		persistedDataHistory, err := ao.accountsq.lookupOnlineHistory(addr)
+		if err != nil {
+			return ledgercore.OnlineAccountData{}, err
 		}
-		ao.accountsMu.RLock()
-		needUnlock = true
-		for currentDbRound >= ao.cachedDBRoundOnline && currentDeltaLen == len(ao.deltas) {
-			ao.accountsReadCond.Wait()
+		err = func() error {
```

Comment:
why are we wrapping this in another function? Just for readability?

Reply:
no, just to use defer for unlocking.